### ISSUE TITLE
[discussion] architecture for clipping/embedded planes and other "geometrical masks"

### ISSUE THREAD

**Issue by brisvag** [napari/napari#3148]
## 🚀 Feature
After a chat with @kevinyamauchi and @alisterburt, we have some ideas for the future of embedded plane rendering (#3023), clipping planes (#3140) and other "geometry-based masks" that modify the rendering (a bounding box, or @alisterburt 's beloved "clipping orb" :P).

## Motivation
Things we are trying to address/implement:
- EventedModels all the way down
- future proofing the architecture to eventually work well with layer groups and layer/data-source decoupling
- allow planes (and possibily other masks) to do multiple things: for example, @kevinyamauchi would like to clip a volume *and also* display an averaged slice on the clipped face. This could be handled by the same plane object. Eventually, this could be done across layers through layer groups (e.g: clip one volume, average a different one on the same plane).

## Pitch

`Layer` holds a `geometrical_masks` (naming is hard, ignore this for now) property. It's an evented list of `Mask` objects (planes, boxes, orbs), or eventually a reference to the `LayerGroup`'s list. This can work like the `LayerList` and allow name-based indexing.
Something like:
```python
>>> layer.geometrical_masks
# [Plane(name='my_plane', position=[0,0,0], normal=[0, 0, 1], thickness=10),
#  Box(name='big_box'...), Orb(...)]
```

A layer can then do stuff with these objects; for example:
```python
layer.clipping.append('big_box')
```
would begin clipping the volume using the box, and:
```python
layer.embedded_plane = 'my_plane'
```
would then start rendering the embedded plane, clipped by any active clipping. The getters should just return a reference to the appropriate `Mask` object.

This becomes particularly useful with `LayerGroups` and the separation between data sources and `Layers`; at that point, one could duplicate layers without doubling memory usage, display them in different ways (clipping/embedded) and still use the same underlying plane object.

It should also make easier to display such objects (without having to pick "which one of these duplicate planes do we show?) in the viewer for ease of interaction (e.g: using the "highlight layer", #2954).

## Alternatives

Another approach would be to hold such planes on the viewer (or canvas?). This would make it easier to
- have a single plane instance controlling several layers 
- hook up callbacks

We found however that this became quickly complicated when it comes to passing around references to layers, especially if one wants to clip different layers in different ways.

The third alternative is to keep it similar to how it is: each of these masks is a different object, and they would instead need to be hooked up to syncronize. This would multiply the number of objects needed if one wants to clip multiple layers. It's also harder maintain a consistent state. Rendering the masks for UI purposes also becomes complicated and messy.

## Additional context

In terms of pratical applications and future proofing, we think this approach is pretty good. However, we're a bit unsure on the higher-level achiterctural planning. Should this list be more generalized and unified (e.g: "rendering-related-stuff"), or more specific and separated (e.g: "planes", "polyhedrons" ...)?

Moreover, we based a lot of our reasoning on not-yet-implemented things like LayerGroups (see #2261) and data/visual decoupling. Are these things moving in the direction we thought?

The work needed to get this up shouldn't be too much (#3140 is basically ready), but it would be nice to hear some thoughts from everythone before I actually do it :)

----
cc @kevinyamauchi @alisterburt

**Comment by sofroniewn** [comment]
Ok nice, this is a good discussion. A couple things I'm taking away are that it's more than just planes we're talking about, they could be other objects. We might want to do more than just clip with these objects.

I think the API you're proposing is a reasonable one, but I can see an alternative framing now which is that a canvas has a list of `geometrical_masks` and operations like `clipping` and then you can assign layers to these.

i.e.

```
viewer.canvas.geometrical_masks
# [Plane(name='my_plane', position=[0,0,0], normal=[0, 0, 1], thickness=10),
#  Box(name='big_box'...), Orb(...)]

# viewer.canvas.clipping.append({'big_box': [my_image_layer, my_labels_layer]})
```

This would allow for our layers to stay simple without knowing clipping information and for these geometrical objects to be reused across layers very easily.

Given this is a major API difference from where we're at right now, I wonder if for PRs like #3140 and #3150 we should keep all the new API private or add the word `experimental` in front of every bit of new API.

I would like to release `0.4.11` asap and a lot of this naming will definitely still be in flight

**Comment by alisterburt** [comment]
I agree with the suggestion that these should live on the canvas/viewer and be easily applied to different layers/purposes. Re: `0.4.11` - an `experimental` tag is a good idea, I think it's worth taking the time to get the API around this stuff figured out, maybe for `0.5.0`? Will post a message in the zulip for visibility

**Comment by brisvag** [comment]
I see the benefit from a high-level api standpoint, but I'm worried that moving these things to the canvas will make the actual rendering code a bit noodly (we still somehow need to get these things to the vispy nodes for each layer, since the clipping happens in the shader...).

Agree 100% on the experimental stuff, so we can keep this discussion going without getting stuck!
